gradio/cli/commands/components/publish.py

```diff
@@ -0,0 +1,167 @@
+
+DOCKERFILE = """
+FROM python:3.9
+
+WORKDIR /code
+
+COPY . .
+
+RUN pip install --no-cache-dir -r requirements.txt
+
+RUN chmod -R 777 /code
+
+ENV PYTHONUNBUFFERED=1 \
+	GRADIO_ALLOW_FLAGGING=never \
+	GRADIO_NUM_PORTS=1 \
+	GRADIO_SERVER_NAME=0.0.0.0 \
+    GRADIO_SERVER_PORT=7860 \
+	SYSTEM=spaces
+
+CMD ["python", "app.py"]
+"""
+
+
+def _publish(
+    wheel_file: Annotated[Path, Argument(help="Path to the wheel directory.")],
+    upload_pypi: Annotated[bool, Option(help="Whether to upload to PyPI.")] = True,
+    pypi_username: Annotated[str, Option(help="The username for PyPI.")] = "",
+    pypi_password: Annotated[str, Option(help="The password for PyPI.")] = "",
+    upload_demo: Annotated[
+        bool, Option(help="Whether to upload demo to HuggingFace.")
+    ] = True,
+    demo_dir: Annotated[
+        Optional[Path], Option(help="Path to the demo directory.")
+    ] = None,
+    hf_token: Annotated[
+        Optional[str],
+        Option(
+            help="HuggingFace token for uploading demo. Can be omitted if already logged in via huggingface cli."
+        ),
+    ] = None,
+):
+    console = Console()
+    wheel_file = wheel_file.resolve()
+    if not wheel_file.suffix == ".whl":
+        raise ValueError(
+            "Please provide a wheel file. It must end with .whl. Run `gradio cc build` to create a wheel file."
+        )
+    if not wheel_file.exists():
+        raise ValueError(
+            f"{wheel_file} does not exist. Run `gradio cc build` to create a wheel file."
+        )
+
+    if upload_pypi and (not pypi_username or not pypi_password):
+        panel = Panel(
+            "You have opted to upload to [blue]https://pypi.org/[/] but no pypi credentials have been provided.\n"
+            "If you do not have an account, register account here: [blue]{PYPI_REGISTER_URL}[/]"
+        )
+        print(panel)
+        pypi_username = Prompt.ask(":laptop_computer: Enter your pypi username")
+        pypi_password = Prompt.ask(":closed_lock_with_key: Enter your pypi password")
+    if upload_pypi:
+        twine_settings = Settings(username=pypi_username, password=pypi_password)
+        try:
+            twine_upload(twine_settings, [str(wheel_file)])
+        except Exception:
+            console.print_exception()
+    if upload_demo and not demo_dir:
+        panel = Panel(
+            "You have opted to upload to [blue]https://huggingface.co/spaces[/] but no demo directory has been provided.\n"
+            "Please provide the path to the [magenta]demo directory[/] for your custom component.\n"
+            "This directory should contain [magenta]all the files[/] it needs to run successfully.\n"
+            "Please make sure the gradio app is in an [magenta]app.py[/] file.\n"
+            "If you need additional python requirements, add a [magenta]requirements.txt[/] file to this directory."
+        )
+        print(panel)
+        demo_dir = Path(Prompt.ask(":hugging_face: Please enter demo directory"))
```

Comment:
if hf_token is None, the hub library will automatically use the one you set during the cli login. If it can't find it, it will raise the appropriate error about not being logged in. So I don't think we need to check ourselves if they are logged in.